packages/financial-templates-lib/src/price-feed/ExpressionPriceFeed.js

```diff
@@ -0,0 +1,117 @@
+// if the user specifies that the Balancer contract is returning non-18 decimal precision prices.
+class ExpressionPriceFeed extends PriceFeedInterface {
+  constructor(priceFeedMap, expression) {
+    super();
+    this.expressionCode = math.parse(expression).compile();
+    this.priceFeedMap = priceFeedMap;
+  }
+
+  async getHistoricalPrice(time) {
+    const historicalPrices = {};
+    const errors = [];
+    await Promise.all(
+      Object.entries(this.priceFeedMap).map(async ([name, pf]) => {
+        const price = await pf.getHistoricalPrice(time).catch(err => errors.push(err));
+        historicalPrices[name] = this._convertToDecimal(price);
+      })
+    );
+
+    if (errors.length > 0) {
+      throw errors;
+    }
+
+    return this._convertToFixed(this.expressionCode.evaluate(historicalPrices));
+  }
+
+  getLastUpdateTime() {
+    const lastUpdateTimes = Object.values(this.priceFeedMap).map(pf => pf.getLastUpdateTime());
+
+    // If any constituents returned an invalid value, bubble it up.
+    if (lastUpdateTimes.some(time => time === null || time === undefined)) {
+      return null;
+    }
+
+    // Take the max.
+    return Math.max(...lastUpdateTimes);
+  }
+
+  getLookback() {
+    const lookbacks = Object.values(this.priceFeedMap).map(priceFeed => priceFeed.getLookback());
+
+    // If any constituents returned an invalid value, bubble it up.
+    if (lookbacks.some(lookback => lookback === undefined || lookback === null)) {
+      return null;
+    }
+
+    // Take the min since the overall lookback will be the min.
+    return Math.min(...lookbacks);
+  }
+
+  getCurrentPrice() {
+    const prices = {};
+    const errors = [];
+    Object.entries(this.priceFeedMap).map(async ([name, pf]) => {
+      try {
+        const price = pf.getCurrentPrice();
+        assert(price !== undefined && price !== null, "Valid price must be returned");
+        prices[name] = this._convertToDecimal(price);
+      } catch (err) {
+        errors.push(err);
+      }
+    });
+
+    if (errors.length > 0) return null;
+
+    return this._convertToFixed(this.expressionCode.evaluate(prices));
+  }
+
+  getPriceFeedDecimals() {
+    const decimalArray = Object.values(this.priceFeedMap).map(pf => pf.getPriceFeedDecimals());
+    const decimalsValue = decimalArray[0];
+    assert(decimalsValue, "Invalid decimals value");
+    assert(
+      decimalArray.every(decimals => decimals === decimalArray[0]),
+      "Constituent price feeds do not have matching decimals"
+    );
+    return decimalsValue;
+  }
+
+  async update() {
+    // Update all constituent price feeds.
+    await Promise.all(Object.values(this.priceFeedMap).map(pf => pf.update()));
+  }
+
+  // Takes a BN fixed point number and converts it to a math.bignumber decimal number that the math library can handle.
+  _convertToDecimal(price) {
+    const decimals = math.bignumber(this.getPriceFeedDecimals());
```

Comment:
so this limits this feed to only outputting at the decimals of the underlying price feed. It might be worth while to set this as a parameter. for example, there could be a feed that is comprised of a number of underlying feeds that use say 6 decimals but the resultant feed is set to 18 decimals. Or, if running on an old emp, it is possible the decimals don't match underlying feeds.

In ether case, do you think it worthwhile to add a constructor param that scales the output according to some `priceFeedDecimals` in the same way we treat other feed types?

Reply:
It's a fair point. Will add.